src/socket.rs

```diff
@@ -0,0 +1,131 @@
+    pub fn peer_addr(&self) -> ::Res<SocketAddr> {
+        match *self {
+            Socket::Udp(ref sock) => sock.peer_addr(),
+            Socket::Tcp(ref sock) => sock.peer_addr(),
+        }
+    }
+
+    /// Set Time To Live value for the underlying socket.
+    pub fn set_ttl(&self, ttl: u32) -> ::Res<()> {
+        match *self {
+            Socket::Udp(ref sock) => sock.set_ttl(ttl),
+            Socket::Tcp(ref sock) => sock.set_ttl(ttl),
+        }
+    }
+
+    /// Retrieve Time To Live value.
+    pub fn ttl(&self) -> ::Res<u32> {
+        match *self {
+            Socket::Udp(ref sock) => sock.ttl(),
+            Socket::Tcp(ref sock) => sock.ttl(),
+        }
+    }
+
+    /// Retrieve last socket error, if one exists.
+    pub fn take_error(&self) -> ::Res<Option<io::Error>> {
+        match *self {
+            Socket::Udp(ref sock) => sock.take_error(),
+            Socket::Tcp(ref sock) => sock.take_error(),
+        }
+    }
+
+    /// Read message from the connected socket. Call this method once socket becomes readable.
+    ///
+    /// # Returns:
+    ///
+    ///   - Ok(Some(data)): data has been successfully read from the socket
+    ///   - Ok(None):       there is not enough data in the socket. Call `read()`
+    ///                     again in the next invocation of the `ready` handler.
+    ///   - Err(error):     there was an error reading from the socket.
+    pub fn read<T: DeserializeOwned + Serialize>(&mut self) -> ::Res<Option<T>> {
+        match *self {
+            Socket::Udp(ref mut sock) => sock.read(),
+            Socket::Tcp(ref mut sock) => sock.read(),
+        }
+    }
+
+    /// Write a message to the connected socket.
+    ///
+    /// # Returns:
+    ///
+    ///   - Ok(true):   the message has been successfully written.
+    ///   - Ok(false):  the message has been queued, but not yet fully written.
+    ///                 will be attempted in the next write schedule.
+    ///   - Err(error): there was an error while writing to the socket.
+    pub fn write<T: Serialize>(&mut self, msg: Option<(T, Priority)>) -> ::Res<bool> {
+        match *self {
+            Socket::Udp(ref mut sock) => sock.write(msg),
+            Socket::Tcp(ref mut sock) => sock.write(msg),
+        }
+    }
+
+    /// Sets linger time for connection based protocols.
+    pub fn set_linger(&self, dur: Option<Duration>) -> ::Res<()> {
+        match *self {
+            Socket::Udp(_) => Ok(()), // do nothing for UDP socket
+            Socket::Tcp(ref sock) => sock.set_linger(dur),
+        }
+    }
+
+    /// Return the wrapped UDP socket or `None`, if it's a different variant.
+    pub fn into_udp(self) -> Option<UdpSock> {
+        match self {
+            Socket::Udp(sock) => Some(sock),
+            Socket::Tcp(_) => None,
+        }
+    }
+
+    /// Return the wrapped TCP socket or `None`, if it's a different variant.
+    pub fn into_tcp(self) -> Option<UdpSock> {
```

Comment:
@povilasb thinking more i'm thinking skip even `into_...()` now :D .. The user can always just do a match themselves so let them to that instead of bloating the API here.

Reply:
So none of these basically.